### PR TITLE
Focus containers

### PR DESCRIPTION
**Pull request by simenbrekken**
The simplest working implementation so far (probably has shortcomings):

- Searches for the closests `|data-focusable]` in a given direction
- Any invalid candidates are omitted from search. E.g. going right won't give you candidates to the left)
- Applies orthogonal weighting. A fancy way of saying that distances in the opposite direction to where we are headed are multiplied by number.

In terms of limiting the number of candidates we could implement some similar concepts from https://github.com/luke-chang/js-spatial-navigation where we further partition the search space by first finding the closest `[data-focus-section]` and then only searching its candidates.

Preview (click a node inside a container to set initial focus): https://tv-focus-experiments-git-focus-container.simen.vercel.app/

Kapture 2020-11-12 at 22 42 44
_There's no limit to the number of containers with their own currently focused node_

### CODE REVIEW

**Comment by popeindustries** [comment]
Good stuff! This seems like a really clean approach (it took me a few minutes before I realised that `data-focus` and `data-section` are just for demo styling).

File: www/index.html [hunk 1]
```diff
@@ -1,8 +1,99 @@
...
+    <script type="module" src="./index.js"></script>
+    <link rel="stylesheet" href="./index.css" />
+  </head>
+  <body>
+    <h1>Front page</h1>
+    <div
+      data-focus
+      data-focus-container
+      title="front-page"
+      data-orientation="vertical"
```
Comment (popeindustries) [inline www/index.html:12]: `data-focus-orientation` or `data-focus-axis` maybe?

File: www/index.js [hunk 2]
```diff
@@ -1,86 +1,171 @@
...
-import { html, render } from '../node_modules/lit-html/lit-html.js';
-
-class ChannelElement extends ReactiveElement {
-  static get reactiveProperties() {
-    return {
-      channel: Object,
-    };
-  }
+// @ts-check
+const verticalOrthogonalWeight = 30;
```
Comment (popeindustries) [inline www/index.js:2]: Are these values layout dependant? Do they need to be configurable?

File: www/index.js [hunk 3]
```diff
@@ -1,86 +1,171 @@
...
-      { id: 'nrk1', name: 'NRK1' },
-      { id: 'nrk2', name: 'NRK2' },
-      { id: 'nrk3', name: 'NRK3' },
-    ];
-
-    return html`<div>
-      <h1>Live page</h1>
-      <tv-channels-list .channels=${channels}></tv-channels-list>
-    </div>`;
+  document.querySelectorAll('[data-focused]').forEach((
```
Comment (popeindustries) [inline www/index.js:149]: During this experimentation phase, it might be interesting to debug how long this searching operation takes

**Comment by popeindustries** [comment]
One disadvantage of this approach is that it can't automatically recover from changes to the DOM that result in `data-focused` being removed or hidden. I wonder, however, if it would make sense to store the client rect and use that to find the next closest candidate (after a manual call to `resetFocused()` or similar)?

**Comment by popeindustries** [comment]
Any thoughts on how we could determine which container is active when multiple containers are used simultaneously? Is it enough to filter each `data-focused` by hidden state (including ancestors)? Here is how jQuery determines if an element is hidden: `!(elem.offsetWidth || elem.offsetHeight || elem.getClientRects().length)`

**Comment by popeindustries** [comment]
Another thought: do we need `data-focusable`? If we take a standards approach, couldn't we just select all natively focusable elements (`[tabindex],a,button:not([disabled]),input:not([disabled]),select:not([disabled]),textarea:not([disabled]`)?

**Review by ingvilddev — APPROVED** [review]

File: www/index.html (hunk 1 above)
Comment (popeindustries) [reply www/index.html:12]: actually, looks like this attribute isn't used at all?

**Pull request closed without merging**